rfcs/0164-bootspecv2.md

```diff
@@ -0,0 +1,300 @@
+initrds as a set, allowing bootloaders to handle a list of initrds or a single
+initrd in cases where multiple initrds are not supported.
+
+[1]: https://uapi-group.org/specifications/specs/boot_loader_specification/#type-1-boot-loader-entry-keys
+[2]: https://github.com/systemd/systemd/blob/main/src/boot/efi/stub.c#L778-L793
+
+## Initrd secrets
+
+Initrd secrets play a crucial role in the NixOS ecosystem, but they have raised
+concerns, including security vulnerabilities (e.g., [GHSA-3rvf-24q2-24ww](https://github.com/NixOS/calamares-nixos-extensions/security/advisories/GHSA-3rvf-24q2-24ww
+)) and
+issues with booting in the Heads platform (e.g., [Issue #1348](https://github.com/linuxboot/heads/issues/1348.
+)).
+
+Initrd secrets serve to protect boot-time secrets from exposure within the Nix
+store. This is achieved by using runtime scripts that append CPIO archives to
+the generation's initrd during bootloader installation. However, the term
+"initrd secrets" can be misleading, as the secrets are plaintext and can be
+accessed from the ESP partition, offering limited confidentiality and no
+integrity.
+
+In practice, initrd secrets are often employed to establish stable fingerprints
+for SSH servers within the initrd, aiding in remote disk decryption on servers.
+To address these issues, this RFC proposes moving away from the "appender
+script" model in Bootspec v1 and instead adopting a hash map format to
+represent secrets and their corresponding values.
+
+This change allows for greater flexibility and enhanced security. For users in
+the NixOS ecosystem relying on systemd and its semantics, the proposal suggests
+offering a systemd-credentials approach for handling initrd secrets,
+potentially enabling encryption using TPM2 if available, or a straightforward
+key if TPM2 is not present. Additionally, this approach provides the means to
+build using the `LoadEncryptedCredential` abstraction.
+
+For those not using systemd, the hash map format offers full flexibility to
+bootloader implementations to determine how to add the secrets, such as
+appending to the initrd or other methods.
+
+**Key Takeaway**: This RFC recommends changing the format of Bootspec v2 to support
+`initrdSecrets` as a hashmap with string keys and values. The key represents a
+name used to organize secrets, and the value denotes an accessible path during
+bootloader installation.
+
+_Example_
+
+```
+"initrdSecrets": {
+   "my-private-key": "/etc/nixos/secrets/wireguard-key"
+}
+```
+
+## Device Tree and Device Tree Overlays
+
+Non-x86 systems often rely on device trees to inform firmware or bootloader about the available devices and hardware support. 
+
+There are two distinct requirements for device trees: 
+
+- generic hardware support, where firmware or bootloader selects the appropriate device tree
+- device-specific support, where a hardcoded device tree is required. 
+
+The latter indicates a potentially problematic, non-upstreamed, or in-development platform.
```

Comment:
This blurb was written while discussing this with @samueldr who well, by their work, is an expert matter in devicetree handling in the Nix ecosystem. 

In my understanding (note that I am not expert matter, and I only know the UEFI side of this), the point I'm trying to make is not the one of bundling or not extra devicetrees but the one of "how do you decide which devicetree to load?"

If your Linux kernel contains all known devicetree, that's awesome news. Though, the problem starts when you don't have the devicetree in the Linux kernel, like I mention it in

> The latter indicates a potentially problematic, non-upstreamed, or in-development platform.

How do you boot a system with a Linux kernel that doesn't know about the right devicetree? I believe @samueldr knows some platforms like those and hence why this is an upstream problem.

The "single devicetree" hardcoding solution is there to offer an escape hatch, not to say this is the way to go.

---

Though, if given those clarifications, what you are saying is that we should always have `FDTDIR` or we should embed all the devicetrees inside the kernel, am I right that you are saying this would cover for all platforms out there in the world, including broken ones?

And therefore, what would be the support of devicetrees in bootspec in that case? Why not just have always the kernel bundle all **known** devicetree? What is the usefulness of having devicetrees expressed in bootspec with what I believe is your proposal — put everything in the kernel, right?

Reply:
(also as part of open questions, I will reput the link here: https://github.com/aarch64-laptops/edk2/tree/dtbloader-app#dtbloader — which I think is very enlightening on the state of DTB I guess.)

Reply:
These are not "DT" requirements, but the *knowledge NixOS has to have* for ***generation-provided DTs***.

*Thinking about data provided by the generation*, either you want the device tree output following the Linux naming semantics, or you want to pick one. There is no in-between that the boot methods allow.

In the situation where you use the platform firmware provided DT, you just configure neither, because ***it's not data from the generation***. It's ambiant data from the platform firmware.

Reminder, bootspec is not describing the hardware you are running things on, **it describes data from a given NixOS generation that may be required for configuring booting**.